app/com/linkedin/drelephant/exceptions/core/ExceptionFingerprintingSpark.java

```diff
@@ -0,0 +1,368 @@
+   */
+  public boolean isUseRestAPI() {
+    return useRestAPI;
+  }
+
+  public void setUseRestAPI(boolean useRestAPI) {
+    this.useRestAPI = useRestAPI;
+  }
+
+  @Override
+  public List<ExceptionInfo> processRawData(AnalyticJob analyticJob) {
+    this.analyticJob = analyticJob;
+    List<ExceptionInfo> exceptions = new ArrayList<ExceptionInfo>();
+
+    processStageLogs(exceptions);
+    //TODO : If there are enough information from stage log then we should not call
+    //driver logs ,to optimize the process .But it can lead to false positivies  in the system,
+    //since failure of stage may or may not be the reason for application failure
+    processDriverLogs(exceptions);
+    return exceptions;
+  }
+
+  /**
+   * process stage logs
+   * @param exceptions
+   */
+  private void processStageLogs(List<ExceptionInfo> exceptions) {
+    long startTime = System.nanoTime();
+    try {
+      if (failedStageData != null && failedStageData.size() > 0) {
+        for (StageData stageData : failedStageData) {
+          /**
+           * Currently there is no use of exception unique ID . But in future it can be used to
+           * find out simillar exceptions . We might need to change from hashcode to some other id
+           * which will give same id , if they similar exceptions .
+           */
+          //TODO : ID should be same for simillar exceptions
+          addExceptions((stageData.failureReason().get() + "" + stageData.details()).hashCode(),
+              stageData.failureReason().get(), stageData.details(), ExceptionSource.EXECUTOR, exceptions);
+        }
+      } else {
+        logger.info(" There is not failed stage data ");
+      }
+    } catch (Exception e) {
+      logger.error("Error process stages logs ", e);
+    }
+    long endTime = System.nanoTime();
+    logger.info(" Total exception/error parsed so far - stage " + exceptions.size());
+    logger.info(" Time taken for processing stage logs " + (endTime - startTime) * 1.0 / (1000000000.0) + "s");
+  }
+
+  private void addExceptions(int uniqueID, String exceptionName, String exceptionStackTrace,
+      ExceptionSource exceptionSource, List<ExceptionInfo> exceptions) {
+    ExceptionInfo exceptionInfo = new ExceptionInfo(uniqueID, exceptionName, exceptionStackTrace, exceptionSource);
+    if (debugEnabled) {
+      logger.info(" Exception Information " + exceptionInfo);
+    }
+    exceptions.add(exceptionInfo);
+  }
+
+  /**
+   * process driver / Application master logs for exception
+   * @param exceptions
+   */
+  //ToDo: In case of unable to get log length there are
+  //multiple calls to same URL (we can optimize this)
+  // This case shouldn't happen in unless there are some changes
+  //in JHS APIs
+  private void processDriverLogs(List<ExceptionInfo> exceptions) {
+    long startTime = System.nanoTime();
+    HttpURLConnection connection = null;
```

Comment:
Common code between processDriverLogs and completeURLToQuery can be moved to a utility method.